feature(server): Bring back inline scheduling

royjacobson opened the pull request:
#908 added a path to schedule transactions immediately if the operator fiber was on the same thread as the shard of the relevant keys.

This caused a replication bug: When the replica would receive a command from the journal it could schedule it immediately, before the matching RDB record was loaded. Since the RDB reader is not using the regular transaction mechanisms, it's tricky to use the regular locking mechanism to fix the issue. Instead of locking, I fixed it by completely disabling inline scheduling when the DB is in a 'LOADING' state.

Close #1036

dranikpg commented:
If possible, please make a benchmark

When we discussed inline scheduling, we actually saw that the time for accessing a thread local has an impact on performance - this is why we store the coordinator index and don't access it with proactorbase::me(). Though we changed thread_local to the simpler thread specifier in some cases (don't remember fully)

adiholden commented on .github/workflows/ci.yml:
```diff
@@ -77,7 +77,7 @@ jobs:
           ninja src/all
           ccache --show-stats
           echo Run ctest -V -L DFLY
-          #GLOG_logtostderr=1 GLOG_vmodule=transaction=1,engine_shard_set=1
```
remove

adiholden commented on src/server/transaction.cc:
```diff
@@ -688,7 +688,17 @@ OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
+    // will be scheduled before RdbLoader::LoadItemsBuffer is finished. We can't use the regular
+    // locking mechanism because RdbLoader is not using transactions.
+    if (coordinator_index_ == unique_shard_id_ &&
+        ServerState::tlocal()->gstate() != GlobalState::LOADING) {
```
It will be more clear if you check !is_master and sync_in_progress. See the code in ServerFamily::Role

royjacobson replied:
I don't think I have a good way to get a reference to the global `Replica` object - it's stored in `ServerFamily` and it's not thread safe to read as well. And if I check `gstate()` then checking `is_master` is a bit redundant, no?

royjacobson commented:
> If possible, please make a benchmark
> 
> When we discussed inline scheduling, we actually saw that the time for accessing a thread local has an impact on performance - this is why we store the coordinator index and don't access it with proactorbase::me() (or was it even some other side effect 🤷🏻‍♂️ ) Though we changed thread_local to the simpler thread specifier in some cases (don't remember fully)

looks good, I think -

[SingleHopBench.txt](https://github.com/dragonflydb/dragonfly/files/11310803/SingleHopBench.txt)

romange commented:
I do not think we can merge this PR yet. See https://github.com/dragonflydb/dragonfly/issues/1036#issuecomment-1509002828

royjacobson commented:
I fixed the other preemption bug. It was pretty hard to even confirm it exists; At the end I managed to create inconsistent replication states by 
1 - purposefully inserting random sleeps into the journal callbacks loop and 
2 - running the attached script ->

[trigger.txt](https://github.com/dragonflydb/dragonfly/files/11435837/trigger.txt)

I don't think that the script is not going into CI anytime soon :sweat_smile: But at least I could check that the problem is real and fixed by the patch.

royjacobson commented:
Also, a nice benefit I didn't realize before - this is pretty helpful when we replicate instances with the same number of shards because we always send the data to the correct thread.

adiholden commented on src/server/server_state.h:
```diff
@@ -206,6 +222,11 @@ class ServerState {  // public struct - to allow initialization.
+  // Inline scheduling is an optimization for running in low thread count situations.
+  // Since it bypasses the regular transactions mechanism, it has to be disabled in
+  // some situations.
+  bool journal_callbacks_active_ = false;
```
Can you move this member to the JournalSlice class and have a fuction to Journal to check if there is journal_callback inside  AllowInlineScheduling

adiholden commented on src/server/server_state.cc:
```diff
@@ -70,6 +71,24 @@ void ServerState::Destroy() {
+  // because then we might interleave the callbacks loop from an inlined-scheduled command
+  // and a normally-scheduled command.
+  // The problematic loop is in JournalSlice::AddLogRecord, going over all the callbacks.
+  if (gstate_ == GlobalState::LOADING)
```
This check is to make sure we are not the replica? This is a bit confusing with the comment

adiholden replied:
just put the relevant comment above the relevant if and add to the comment if it reference to the replica side or master side

adiholden commented on src/server/journal/journal_slice.h:
```diff
@@ -44,6 +44,9 @@ class JournalSlice {
   uint32_t RegisterOnChange(ChangeCallback cb);
   void UnregisterOnChange(uint32_t);
+  bool HasRegisteredCallbacks() const {
+    return !change_cb_arr_.empty();
```
dont we need the lock here?

chakaz replied:
Might I suggest `ABSL_GUARDED_BY` for these cases exactly?
Abseil provides thread annotations, enforced by (at least) clang. You can read more about them here: https://abseil.io/docs/cpp/guides/synchronization#thread-annotations
I just used one in my latest PR.

royjacobson replied:
Done. I'm not super happy about adding a lock there, but I did profiling and it was something like 0.1% overhead so I'm not going to optimize because of that.

adiholden commented:
@royjacobson just make sure you merge only after Roman creates a tag for the new version
